recirq/quantum_chess/circuit_transformer.py

```diff
@@ -305,6 +307,391 @@ def transform(self, circuit: cirq.Circuit) -> cirq.Circuit:
         return circuit.transform_qubits(lambda q: self.mapping[q])
 
 
+class DynamicLookAheadHeuristicCircuitTransformer(CircuitTransformer):
+    """Optimizer that transforms a circuit to satify a device's constraints.
+
+    This implements the initial mapping algorithm and the SWAP update algorithm
+    proposed by the paper "A Dynamic Look-Ahead Heuristic for the Qubit Mapping
+    Problem of NISQ Computer:
+    https://ieeexplore.ieee.org/abstract/document/8976109.
+
+    Reference:
+    P. Zhu, Z. Guan and X. Cheng, "A Dynamic Look-Ahead Heuristic for the
+    Qubit Mapping Problem of NISQ Computers," in IEEE Transactions on Computer-
+    Aided Design of Integrated Circuits and Systems, vol. 39, no. 12, pp. 4721-
+    4735, Dec. 2020, doi: 10.1109/TCAD.2020.2970594.
+    """
+    def __init__(self, device: cirq.Device):
+        super().__init__()
+        self.device = device
+
+    def build_physical_qubits_graph(
+        self,
+    ) -> Dict[cirq.GridQubit, List[cirq.GridQubit]]:
+        """Returns an adjacency graph of physical qubits of the device.
+
+        Each edge is bidirectional, and represents a valid two-qubit gate.
+        """
+        g = defaultdict(list)
+        for q in self.device.qubit_set():
+            neighbors = [n for n in q.neighbors() if n in self.device.qubit_set()]
+            for neighbor in neighbors:
+                if q not in g[neighbor]:
+                    g[neighbor].append(q)
+        return g
+
+    def get_least_connected_qubit(
+        self,
+        g: Dict[cirq.Qid, List[Tuple[cirq.Qid, int]]],
+        component: Deque[cirq.Qid],
+    ) -> cirq.Qid:
+        """Returns the least connected qubit.
+
+        Args:
+          g: A logical qubits graph.
+          component: A deque of qubits belonging to the same component.
+        """
+        qubit = None
```

Comment:
if you want, you can use min() with an ordering function to compute this directly:
```
return min(component, key=lambda q: len(g[q]))
```

Reply:
nice, updated!